{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span id=\"top\"></span>\n",
    "\n",
    "#### Final_Project_1 (Score: 20.0 / 20.0)\n",
    "\n",
    "1.  [Written response](#cell-105665040cea53f2) (Score: 3.0 / 3.0)\n",
    "2.  [Written response](#cell-a32f7bee88778be3) (Score: 3.0 / 3.0)\n",
    "3.  [Written response](#cell-69d3e5d10151be16) (Score: 2.0 / 2.0)\n",
    "4.  [Written response](#cell-cb035fc1caba4917) (Score: 4.0 / 2.0)\n",
    "5.  [Written response](#cell-b61be3d778e29197) (Score: 0.0 / 2.0)\n",
    "6.  [Written response](#cell-9cb1df2a39c4b46a) (Score: 3.0 / 3.0)\n",
    "7.  [Comment](#comment-cell-9cb1df2a39c4b46a)\n",
    "8.  [Coding free-response](#cell-575e76cef8940133) (Score: 0.25 / 0.25)\n",
    "9.  [Coding free-response](#cell-dce89a72292cde8e) (Score: 0.25 / 0.25)\n",
    "10. [Coding free-response](#cell-8581042beb51b4fc) (Score: 1.0 / 1.0)\n",
    "11. [Coding free-response](#cell-779a2106cd2beb42) (Score: 3.25 / 3.25)\n",
    "12. [Coding free-response](#cell-ab43aa2854f430a8) (Score: 0.25 / 0.25)\n",
    "\n",
    "### Honors Computational Genetics<a href=\"#Honors-Computational-Genetics\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "#### Final Project Step \\#1: Summarize Data and Assess Quality<a href=\"#Final-Project-Step-#1:-Summarize-Data-and-Assess-Quality\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "###### As with most Genomic projects, the first step in this project involves summarizing your data and filtering it to exclude low-quality sequences. You will need to:<a href=\"#As-with-most-Genomic-projects,-the-first-step-in-this-project-involves-summarizing-your-data-and-filtering-it-to-exclude-low-quality-sequences.-You-will-need-to:\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "-   Calculate Pre-filtering Summary Information for your data\n",
    "-   Filter your data using `trimmomatic`\n",
    "-   Calculate Post-filtering Summary Information for the filtered daat\n",
    "\n",
    "When running `trimmomatic`, be sure to specify options that will have it\n",
    "do the following:\n",
    "\n",
    "1.  Only keep reads if they are at least 50 bp in length after trimming\n",
    "2.  Remove the first nucleotide from the sequence if it has a Phred\n",
    "    quality score between 0 and 10, and continue removing subsequent\n",
    "    nucleotides until reaching one with a quality score above 10\n",
    "3.  Same as \\#2 but for the nucleotides at the end of the sequence\n",
    "4.  Calculate the average quality score for consecutive nucleotides in a\n",
    "    window of size 4. If the score drops below 15, the remaining\n",
    "    sequence should be trimmed.\n",
    "5.  Include the same option from the lecture notebook that tells\n",
    "    Trimmomatic to clip the adapter sequences.\n",
    "\n",
    "###### See the Results section for the summary information you will need to collect before and after trimming<a href=\"#See-the-Results-section-for-the-summary-information-you-will-need-to-collect-before-and-after-trimming\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "### Question<a href=\"#Question\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "*Write 3-6 sentences that provide background information about this\n",
    "study and describe the main question and/or goal of the project.*\n",
    "\n",
    "*Be sure to address the following topics:*\n",
    "\n",
    "-   What species is being studied?\n",
    "-   Where were the individuals collected from? Were they kept in the lab\n",
    "    before being sequenced?\n",
    "-   What type of sequencing was performed? Paired-end or single-end? How\n",
    "    long are the sequences?\n",
    "-   Were any other analyses done with the sequences before they were\n",
    "    given to you?\n",
    "\n",
    "<span id=\"comment-cell-105665040cea53f2\"></span><span\n",
    "id=\"cell-105665040cea53f2\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 3.0 / 3.0 [(Top)](#top) </span>\n",
    "\n",
    "The species being studied is the Drosophila simulans, from which a\n",
    "sample of the flies was captured in Zuma, California and then inbred for\n",
    "several generations in a lab. Single-end illumina sequencing was\n",
    "performed for each of the flies and these sequences, which are billions\n",
    "of basepairs in length, were aligned to a reference genome for the D.\n",
    "simulans species. While most of the sequences aligned to the reference\n",
    "genome, there were some that did not, meaning that they are not usually\n",
    "a part of the D. simulans genome. Thus, the goal of this project is to\n",
    "determine where these unaligned sequences came from.\n",
    "\n",
    "### Hypothesis<a href=\"#Hypothesis\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "*Write 1-3 sentences describing the possible explanations for why your\n",
    "Illumina sequences did not align to the D. simulans reference genome.\n",
    "Include a prediction of which explanation is most important (i.e. which\n",
    "one will apply to the majority of sequences)*\n",
    "\n",
    "<span id=\"comment-cell-a32f7bee88778be3\"></span><span\n",
    "id=\"cell-a32f7bee88778be3\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 3.0 / 3.0 [(Top)](#top) </span>\n",
    "\n",
    "I predict that the unaligned sequences were caused by viruses that were\n",
    "present in only some of the individual flies because the sequences had\n",
    "relatively high quality scores before and after filtering, making it\n",
    "unlikely that these sequences were a result of a sequencing error. I\n",
    "believe that these unaligned sequences are not likely from microbes in\n",
    "the flies because the entire sample was fed the same strain of yeast and\n",
    "since Wolbachia is known to prevent the birth of male flies, inbreeeding\n",
    "the flies would have been difficult.\n",
    "\n",
    "### Methods<a href=\"#Methods\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "*The paragraph below should briefly describe how you were able to\n",
    "calculate the summary information about your Illumina data that is\n",
    "presented in the Results section*\n",
    "\n",
    "<span id=\"comment-cell-69d3e5d10151be16\"></span><span\n",
    "id=\"cell-69d3e5d10151be16\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 2.0 / 2.0 [(Top)](#top) </span>\n",
    "\n",
    "To obtain the summary information, I read in the compressed fastq files\n",
    "using the gzip module and the SeqIO module from Biopython. Within the\n",
    "loop, I counted the number of sequences there are in the file by\n",
    "accumulating the number of loops that ran and to find the total number\n",
    "of basepairs, I accumulated the length of each sequence in the file.\n",
    "Next, I generated a dictionary of the quality scores for each nucleotide\n",
    "in each sequence and used indexing to work with the first four and last\n",
    "four nucleotides in each sequence. I added the scores of the first four\n",
    "together and, separately, the last four together and divided by four to\n",
    "get the averages in each sequence. Next, I took the sum of all of the\n",
    "beginning average quality scores and, again separately, the sum of all\n",
    "of the final average quality scores to calculate the overall average\n",
    "scores for the first four and last four nucleotides of the file.\n",
    "Finally, I averaged the quality scores of every nucleotide in each\n",
    "sequence and summed up those values to calculate the overall quality\n",
    "score of the entire file.\n",
    "\n",
    "*This paragraph should briefly describe how you ran Trimmomatic\n",
    "including descriptions of the parameters you used and what they mean.*\n",
    "\n",
    "<span id=\"comment-cell-cb035fc1caba4917\"></span><span\n",
    "id=\"cell-cb035fc1caba4917\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 4.0 / 2.0 [(Top)](#top) </span>\n",
    "\n",
    "I used Trimmomatic by first importing the java and Trimmomatic modules.\n",
    "Then, I ran the Trimmomatic command in java using the parameters noted\n",
    "above to cut out the nucleotides with low quality scores from the\n",
    "sequences and the resulting filtered sequences were saved into a new\n",
    "fastq file.\n",
    "\n",
    "### Results<a href=\"#Results\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "|                              | Pre-filtering      | Post-filtering    |\n",
    "|------------------------------|--------------------|-------------------|\n",
    "| Total \\# of Sequences        | 4312924            | 4138200           |\n",
    "| Total \\# of Basepairs        | 474421640          | 412688981         |\n",
    "| Avg. Quality of First 4 nucl | 33.688164630306495 | 32.36462374481906 |\n",
    "| Avg. Quality of Last 4 nucl  | 33.110935295868885 | 33.70855143749345 |\n",
    "| Avg. Quality of Sequence     | 36.424134350173475 | 35.83045381600684 |\n",
    "\n",
    "<span id=\"comment-cell-b61be3d778e29197\"></span><span\n",
    "id=\"cell-b61be3d778e29197\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 0.0 / 2.0 [(Top)](#top) </span>\n",
    "\n",
    "*Once the table is complete, delete the instructions under the Results\n",
    "header and write a few sentences below summarizing your table. You\n",
    "should address how the quality changed after filtering your data and\n",
    "include the percentage of the orginal data that you were able to keep\n",
    "after filtering.*\n",
    "\n",
    "<span id=\"comment-cell-9cb1df2a39c4b46a\"></span><span\n",
    "id=\"cell-9cb1df2a39c4b46a\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 3.0 / 3.0 [(Top)](#top) </span>\n",
    "\n",
    "There was an unusual change in the quality scores of the sequences after\n",
    "filtering the unaligned data. Normally, the quality score of a sequence\n",
    "is expected to increase after filtering it because the nucleotides with\n",
    "low quality scores are removed. However, in my analyis, while the\n",
    "average quality score of the last four nucleotides increased slightly,\n",
    "the other average scores decreased. After talking to Dr. Ellison about\n",
    "this, although it is definitely unusual, it could be because the\n",
    "unfiltered sequences already have relatively high quality scores and\n",
    "most of the sequences, 95.05%, were kept after filtering.\n",
    "\n",
    "**Comments:**\n",
    "\n",
    "Very well done, handling some unusual results.\n",
    "\n",
    "### Code<a href=\"#Code\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "###### Here is where you should include the code and Unix commands that you used for the analyses that are summarized in the Results section<a href=\"#Here-is-where-you-should-include-the-code-and-Unix-commands-that-you-used-for-the-analyses-that-are-summarized-in-the-Results-section\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "*Copy your Illumina data to your current directory, replacing NETID with\n",
    "your own NetID:*\n",
    "\n",
    "`/projects/genetics/genetics203/data/final_project/NETID.unaligned.fastq.gz`\n",
    "\n",
    "In \\[9\\]:\n",
    "\n",
    "<span id=\"comment-cell-575e76cef8940133\"></span><span\n",
    "id=\"cell-575e76cef8940133\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 0.25 / 0.25 [(Top)](#top) </span>\n",
    "\n",
    "    cp /projects/genetics/genetics203/data/final_project/hi70.unaligned.fastq.gz .\n",
    "\n",
    "*Load the modules that you will need to run Trimmomatic*\n",
    "\n",
    "In \\[10\\]:\n",
    "\n",
    "<span id=\"comment-cell-dce89a72292cde8e\"></span><span\n",
    "id=\"cell-dce89a72292cde8e\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 0.25 / 0.25 [(Top)](#top) </span>\n",
    "\n",
    "    %%bash\n",
    "    module use /projects/community/modulefiles\n",
    "    module load java\n",
    "    module load Trimmomatic\n",
    "\n",
    "*Your Trimmomatic command should be in the cell below*\n",
    "\n",
    "**IMPORTANT: you are working with millions of sequences, so this command\n",
    "will take 5-10 minutes to run. Do NOT try to run any other commands\n",
    "until it finishes**\n",
    "\n",
    "In \\[1\\]:\n",
    "\n",
    "<span id=\"comment-cell-8581042beb51b4fc\"></span><span\n",
    "id=\"cell-8581042beb51b4fc\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 1.0 / 1.0 [(Top)](#top) </span>\n",
    "\n",
    "    java -Xmx1G -jar /projects/community/Trimmomatic/Trimmomatic-0.38/trimmomatic-0.38.jar SE -threads 1 hi70.unaligned.fastq.gz hi70.unaligned_trimmed.fastq.gz LEADING:10 TRAILING:10 SLIDINGWINDOW:4:15 ILLUMINACLIP:TruSeq3-SE.fa:2:30:10 MINLEN:50\n",
    "\n",
    "    TrimmomaticSE: Started with arguments:\n",
    "     -threads 1 hi70.unaligned.fastq.gz hi70.unaligned_trimmed.fastq.gz LEADING:10 TRAILING:10 SLIDINGWINDOW:4:15 ILLUMINACLIP:TruSeq3-SE.fa:2:30:10 MINLEN:50\n",
    "    Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
    "    Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
    "    ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
    "    Quality encoding detected as phred33\n",
    "    Input Reads: 4312924 Surviving: 4138200 (95.95%) Dropped: 174724 (4.05%)\n",
    "    TrimmomaticSE: Completed successfully\n",
    "\n",
    "*Your python code should be in the cell below*\n",
    "\n",
    "In \\[2\\]:\n",
    "\n",
    "<span id=\"comment-cell-779a2106cd2beb42\"></span><span\n",
    "id=\"cell-779a2106cd2beb42\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 3.25 / 3.25 [(Top)](#top) </span>\n",
    "\n",
    "    import gzip\n",
    "    from Bio import SeqIO\n",
    "\n",
    "    original_seq=0              # number of sequences\n",
    "    original_bp=0               # number of basepairs\n",
    "    original_seq_qual=0         # average quality score for entire sequence\n",
    "    sum_first_avg=0             # sum of the averages of the first 4 bases of every sequence\n",
    "    sum_last_avg=0              # sum of the averages of the last 4 bases of every sequence\n",
    "    sum_avg=0                   # sum of averages of each sequence\n",
    "    original_first_avg=0        # average quality score for first 4 bases\n",
    "    original_last_avg=0         # average quality score for last 4 bases\n",
    "\n",
    "    fh=gzip.open('hi70.unaligned.fastq.gz', 'rt')\n",
    "\n",
    "    for record in SeqIO.parse(fh, 'fastq'):\n",
    "        original_seq+=1\n",
    "        num_bp=len(record)\n",
    "        original_bp+=num_bp\n",
    "        qlist=record.letter_annotations['phred_quality']\n",
    "        sum_first=0\n",
    "        for i in range(4):\n",
    "            sum_first+=int(qlist[i])\n",
    "        avg_first_seq=sum_first/4\n",
    "        sum_first_avg+=avg_first_seq\n",
    "        sum_last=0\n",
    "        for i in range(-4,0):\n",
    "            sum_last+=int(qlist[i])\n",
    "        avg_last_seq=sum_last/4\n",
    "        sum_last_avg+=avg_last_seq\n",
    "        seq_qual=0\n",
    "        for value in qlist:\n",
    "            seq_qual+=int(value)\n",
    "        average=seq_qual/len(record)\n",
    "        sum_avg+=average\n",
    "\n",
    "    original_first_avg=sum_first_avg/original_seq\n",
    "    original_last_avg=sum_last_avg/original_seq\n",
    "    original_seq_qual=sum_avg/original_seq\n",
    "\n",
    "    print(original_seq)\n",
    "    print(original_bp)\n",
    "    print(original_first_avg)\n",
    "    print(original_last_avg)\n",
    "    print(original_seq_qual)\n",
    "\n",
    "    4312924\n",
    "    474421640\n",
    "    33.688164630306495\n",
    "    33.110935295868885\n",
    "    36.424134350173475\n",
    "\n",
    "In \\[3\\]:\n",
    "\n",
    "    trim_seq=0            \n",
    "    trim_bp=0       \n",
    "    trim_seq_qual=0        \n",
    "    sum_first_avg=0          \n",
    "    sum_last_avg=0           \n",
    "    sum_avg=0                   \n",
    "    trim_first_avg=0     \n",
    "    trim_last_avg=0         \n",
    "\n",
    "    fh=gzip.open('hi70.unaligned_trimmed.fastq.gz', 'rt')\n",
    "\n",
    "    for record in SeqIO.parse(fh, 'fastq'):\n",
    "        trim_seq+=1\n",
    "        num_bp=len(record)\n",
    "        trim_bp+=num_bp\n",
    "        qlist=record.letter_annotations['phred_quality']\n",
    "        sum_first=0\n",
    "        for i in range(4):\n",
    "            sum_first+=int(qlist[i])\n",
    "        avg_first_seq=sum_first/4\n",
    "        sum_first_avg+=avg_first_seq\n",
    "        sum_last=0\n",
    "        for i in range(-4,0):\n",
    "            sum_last+=int(qlist[i])\n",
    "        avg_last_seq=sum_last/4\n",
    "        sum_last_avg+=avg_last_seq\n",
    "        seq_qual=0\n",
    "        for value in qlist:\n",
    "            seq_qual+=int(value)\n",
    "        average=seq_qual/len(record)\n",
    "        sum_avg+=average\n",
    "\n",
    "    trim_first_avg=sum_first_avg/original_seq\n",
    "    trim_last_avg=sum_last_avg/original_seq\n",
    "    trim_seq_qual=sum_avg/original_seq\n",
    "\n",
    "    print(trim_seq)\n",
    "    print(trim_bp)\n",
    "    print(trim_first_avg)\n",
    "    print(trim_last_avg)\n",
    "    print(trim_seq_qual)\n",
    "\n",
    "    4138200\n",
    "    412688981\n",
    "    32.36462374481906\n",
    "    33.70855143749345\n",
    "    35.83045381600684\n",
    "\n",
    "##### IMPORTANT: when you are finished, delete the original FASTQ file since you do not need it any more, but KEEP your trimmed file<a href=\"#IMPORTANT:-when-you-are-finished,-delete-the-original-FASTQ-file-since-you-do-not-need-it-any-more,-but-KEEP-your-trimmed-file\" class=\"anchor-link\">¶</a>\n",
    "\n",
    "In \\[5\\]:\n",
    "\n",
    "<span id=\"comment-cell-ab43aa2854f430a8\"></span><span\n",
    "id=\"cell-ab43aa2854f430a8\"></span>\n",
    "\n",
    "<span class=\"nbgrader-label\">Student's answer</span><span\n",
    "class=\"pull-right\"> Score: 0.25 / 0.25 [(Top)](#top) </span>\n",
    "\n",
    "    %%bash\n",
    "    rm hi70.unaligned.fastq.gz\n",
    "\n",
    "In \\[4\\]:\n",
    "\n",
    "    %%bash\n",
    "    zcat hi70.unaligned.fastq.gz | head -n 400 > test.unaligned.fastq\n",
    "\n",
    "In \\[5\\]:\n",
    "\n",
    "    %%bash\n",
    "    gzip test.unaligned.fastq"
   ]
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
